test/lib/services/html_converter_test.rb

```diff
@@ -199,26 +288,79 @@ def test_inject_lang_html_tag_with_lang_in_html_tag
         Wovnrb.get_env('url' => 'http://my-site.com/'),
         Wovnrb.get_settings(settings)
       )
-      converter = HtmlConverter.new(get_dom('<html lang="th"><body>hello</body></html>'), store, headers)
+      url_lang_switcher = Wovnrb::UrlLanguageSwitcher.new(store)
+      converter = HtmlConverter.new(get_dom('<html lang="th"><body>hello</body></html>'), store, headers, url_lang_switcher)
       converter.send(:inject_lang_html_tag)
       expected_html = '<html lang="th"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"></head><body>hello</body></html>'
       assert_equal(expected_html, converter.send(:html))
     end
 
+    def test_translate_canonical_tag
+      settings = default_store_settings
+      store = Wovnrb::Store.instance
+      store.update_settings(settings)
+
+      headers = Wovnrb::Headers.new(
+        Wovnrb.get_env('url' => 'http://my-site.com/?wovn=fr'),
+        store.settings
+      )
+      url_lang_switcher = Wovnrb::UrlLanguageSwitcher.new(store)
+      converter = HtmlConverter.new(get_dom('<html lang="th"><head><link rel="canonical" href="http://my-site.com" /></head><body>hello</body></html>'), store, headers, url_lang_switcher)
+      converter.send(:translate_canonical_tag)
+      expected_html = '<html lang="th"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><link rel="canonical" href="http://my-site.com?wovn=fr"></head><body>hello</body></html>'
+      assert_equal(expected_html, converter.send(:html))
+    end
+
+    def test_translate_canonical_tag_path_pattern
+      settings = default_store_settings
+      settings['url_pattern'] = 'path'
+      store = Wovnrb::Store.instance
+      store.update_settings(settings)
+
+      headers = Wovnrb::Headers.new(
+        Wovnrb.get_env('url' => 'http://my-site.com/fr/'),
+        store.settings
+      )
+      url_lang_switcher = Wovnrb::UrlLanguageSwitcher.new(store)
+      converter = HtmlConverter.new(get_dom('<html lang="th"><head><link rel="canonical" href="http://my-site.com/" /></head><body>hello</body></html>'), store, headers, url_lang_switcher)
+      converter.send(:translate_canonical_tag)
```

Comment:
in the same file we're already testing this with public methods. do we really need to test private methods?

Reply:
Strictly speaking we dont. I first wrote those private method tests and were satisfied with them. But then I realized it's not actually testing the option switch, so I added more tests on the `build` method. Personally I want to keep them because it's easier to see what's going on without the long snippets in the way.